core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala

```diff
@@ -1429,47 +1453,59 @@ class TimeSeriesShard(val ref: DatasetRef,
-    *                          startTimes from Lucene are included
-    * @param partIdsNotInMemory is a collection of partIds fully not in memory
-    */
-  case class IterationResult(partsInMemoryIter: PartitionIterator,
-                             partIdsInMemoryMayNeedOdp: debox.Map[Int, Long] = debox.Map.empty,
-                             partIdsNotInMemory: debox.Buffer[Int] = debox.Buffer.empty)
+  protected def schemaIDFromPartID(partID: Int): Int = {
+    partitions.get(partID) match {
+      case TimeSeriesShard.OutOfMemPartition =>
+        partKeyIndex.partKeyFromPartId(partID).map { pkBytesRef =>
+          val unsafeKeyOffset = PartKeyLuceneIndex.bytesRefToUnsafeOffset(pkBytesRef.offset)
+          RecordSchema.schemaID(pkBytesRef.bytes, unsafeKeyOffset)
+        }.getOrElse(-1)
+      case p: TimeSeriesPartition => p.schema.schemaHash
+    }
+  }
 
   /**
-    * See documentation for IterationResult.
+    * Looks up partitions and schema info from ScanMethods, usually by doing a Lucene search.
+    * Also returns detailed information about what is in memory and not, and does schema discovery.
     */
-  def iteratePartitions(partMethod: PartitionScanMethod,
-                        chunkMethod: ChunkScanMethod): IterationResult = partMethod match {
-    case SinglePartitionScan(partition, _) => IterationResult(ByteKeysPartitionIterator(Seq(partition)))
-    case MultiPartitionScan(partKeys, _)   => IterationResult(ByteKeysPartitionIterator(partKeys))
+  def lookupPartitions(partMethod: PartitionScanMethod,
+                       chunkMethod: ChunkScanMethod): PartLookupResult = partMethod match {
+    case SinglePartitionScan(partition, _) =>
+      PartLookupResult(shardNum, chunkMethod, ByteKeysPartitionIterator(Seq(partition)),
+                       Some(RecordSchema.schemaID(partition)))
+    case MultiPartitionScan(partKeys, _)   =>
+      PartLookupResult(shardNum, chunkMethod, ByteKeysPartitionIterator(partKeys),
+                       partKeys.headOption.map(RecordSchema.schemaID))
     case FilteredPartitionScan(split, filters) =>
-      // TODO: There are other filters that need to be added and translated to Lucene queries
-      if (filters.nonEmpty) {
-        val indexIt = partKeyIndex.partIdsFromFilters(filters, chunkMethod.startTime, chunkMethod.endTime)
-        IterationResult(new InMemPartitionIterator(indexIt))
-      } else {
-        IterationResult(PartitionIterator.fromPartIt(partitions.values.iterator.asScala))
+      // No matter if there are filters or not, need to run things through Lucene so we can discover potential
+      // TSPartitions to read back from disk
+      val coll = partKeyIndex.partIdsFromFilters2(filters, chunkMethod.startTime, chunkMethod.endTime)
+
+      // first find out which partitions are being queried for data not in memory
+      val _schema = Option(coll.getFirstSetBit).filter(_ >= 0).map(schemaIDFromPartID)
```

Comment:
I don't think there is, unless we pre-lookup all the possible schemas from Lucene.  However we decided against that approach.
However if more than one schema exists in the cluster for those filters, it will result in an error  either way, so the end result to the user is deterministic.

Reply:
sounds reasonable.